src/gwt/src/org/rstudio/core/client/ElementIds.java

```diff
@@ -21,19 +21,30 @@
 
 public class ElementIds
 {
-   public static void assignElementId(Element ele, String id)
+   /**
+    * Return a unique ID based on examination of existing elements. Must be assigned
+    * immediately to an element in the DOM to ensure uniqueness.
+    * @param baseId
+    * @return
+    */
+   public static String getUniqueElementId(String baseId)
    {
-      String elementIdBase = ID_PREFIX + id;
+      String elementIdBase = getElementId(baseId);
       String elementId = elementIdBase;
       int counter = 0;
-      
+
       // ensure uniqueness; for example, if multiple modal dialogs are displayed, make sure
       // the OK button instances, etc., are uniquely identified
       while (DomUtils.getElementById(elementId) != null)
       {
          elementId = elementIdBase + "_" + counter++;
       }
```

Comment:
We want these to be deterministic, for test automation authoring purposes. So in a given scenario (bring up this dialog, then that pane, etc.) want controls to end up with the same ID. Agree it seems ugly and inefficient. Alternative, at the moment, is to define and assign ID constants. There's also the way GWT does it, which is a global counter, so things end up as gwt-uid-42, and so on. That is somewhat deterministic, but even more fragile than the current approach, in that any change in the UI is likely to ripple through a bunch of downstream IDs, whereas having them somewhat namespaced (`DIALOG_OK_BUTTON_02`) makes it a bit less fragile. I'm open to ideas!

Reply:
I suppose we could use a GUID by default, then fall back on assigning constant IDs when automation needs them.

Reply:
Mocking it be deterministic during automation would probably be ideal but as it would likely require a lot of work to build that infrastructure we can also cross that bridge when we come to it.  This is almost definitely not a performance bottleneck right now but I could see it happening in the right circumstance, just good to be aware of it moving forward.

An easier solution is probably to just change the automation framework's selectors. Since ordering matters with this style already it's probably no less fragile to change a `#uid-42` selector to `nth-child` selectors.

Reply:
FWIW, there are no current cases in the automation where ordering matters (very much a theoretical problem at this point), just doing this check to prevent duplicate IDs to avoid accessibility violations in a small number of cases where it happens.

It's not ideal for the automation to use selectors based on the structure of the HTML (nth-child, etc) as that is definitely more fragile and ties the automation to implemention of visual layout (e.g. if/when we start moving away from table-based layout such selectors would probably require significant reworking). Having automation that clicks a specific button is less likely to be impacted by changes such as that. For example, from the actual automation code, better to have:

```
 buildDropDown = (By.ID, 'rstudio_label_build_menu')
```
than
```
 # TODO replace with better selectors
 newSessionBtn = (By.XPATH, '//*[@id="sessionsList"]/div[1]/div[2]/button/span')
```

Reply:
Oh I definitely agree that nth-child is horrifically bad, just saying it'd be no worse than if it relied on specific IDs that we assigned by order. If we can force the automation to use smart selectors we'll be good on all fronts.